### ISSUE TITLE
Bugs due to recent updates

### ISSUE THREAD

**Issue by PeterKottas** [react-page/react-page#749]
Hey @macrozone , I see you are doing quite a lot of good work on the lib! Very happy to see that. While updating to the latest version, I've noticed a few small issues and since I am not currently set up with code, I thought I'd run it past you and you can fix in one of your many releases :) 

It all has to do with background plugin:
Here https://github.com/react-page/react-page/blob/master/packages/plugins/layout/background/src/Controls/BackgroundDefaultControls.tsx#L147 it should be changed from "onDragEnd" to "onChangeCommit", same here https://github.com/react-page/react-page/blob/master/packages/plugins/layout/background/src/Controls/BackgroundDefaultControls.tsx#L127.

Then here: https://github.com/react-page/react-page/blob/master/packages/plugins/layout/background/src/Component/index.tsx#L120 the state should be passed in along with the props, otherwise the preview functionality for darken and lighten stops working. This was previously handled by rendering the renderer in Controls like so: https://github.com/react-page/react-page/blob/857b80ac769af74fe8145ba3e19c05bebc8ae541/packages/plugins/layout/background/src/Controls/BackgroundDefaultControls.tsx#L56. But at some point, it was changed thus breaking the preview functionality. 

Carry on with the great work! I especially love the react-dnd update, that was kicking my ass for a while. Cheers

**Comment by PeterKottas** [comment]
Also when doing all the way to 0.16.0-beta.5, when clicking on background/image plugin in edit mode, I see 
```
Could not find editable: undefined
```
Which happens over here https://github.com/react-page/react-page/blob/1d3393912927aa9c3d0a5647148bb964b7b71a6b/packages/core/src/selector/editable/index.ts#L101.
Not sure why that is. 
Lastly, the draft feature is amazing!!!! But sadly it doesn't work for me. I can still see the "draft" content when rendered via HTMLRenderer. Sorry I am polluting the issue, just wanted to give you the feedback.

**Comment by macrozone** [comment]
@PeterKottas thx so much for your feedback! will fix it!

> Lastly, the draft feature is amazing!!!! But sadly it doesn't work for me. I can still see the "draft" content when rendered via HTMLRenderer. Sorry I am polluting the issue, just wanted to give you the feedback.

I also realized that too and i am just about to debug it. I forgot that the serializer removes some properties and i had to transfer the `isDraft` property over there.

**Comment by macrozone** [comment]
@PeterKottas ok, the Slider bug is because of the upgrade to material-ui 4. i fixed it and will trigger a release

**Comment by macrozone** [comment]
@PeterKottas the second issue is because of this:

My goal was to unify the plugins to this pattern: 

`<><Renderer />{ !readOnly ? <Controls /> : null}</>`

so that the renderer would not need to re-render. however, this did not work with plugin, that used some preview-state in the renderer. e.g. the image-plugin. In these cases, i did:

`<>{ !readOnly ? <Controls /> : <Renderer />}</>`

and then the Controls would need to include the Renderer as well.

I will now do the same in the background-plugin, ist just forgot to check that.

Having said that, we should probably revisit that. I mitigate the headache around these concerns a bit, because i only use custom plugins with `@react-page/create-plugin-materialui` and non of the built-in plugins. Would be cool to migrate the built-in plugins to use that (or a more general) factory-function.

**Comment by macrozone** [comment]
@PeterKottas i could not reproduce `Could not find editable: undefined` but i think i know where it could come from... and i think its already fixed.

can you tell me which version did you use?

0.16.0-beta.7 is been baked now and should fix all issues hopefully

**Comment by PeterKottas** [comment]
> @PeterKottas ok, the Slider bug is because of the upgrade to material-ui 4. i fixed it and will trigger a release

Awesome!

> @PeterKottas the second issue is because of this:
> 
> My goal was to unify the plugins to this pattern:
> 
> `<><Renderer />{ !readOnly ? <Controls /> : null}</>`
> 
> so that the renderer would not need to re-render. however, this did not work with plugin, that used some preview-state in the renderer. e.g. the image-plugin. In these cases, i did:
> 
> `<>{ !readOnly ? <Controls /> : <Renderer />}</>`
> 
> and then the Controls would need to include the Renderer as well.
> 
> I will now do the same in the background-plugin, ist just forgot to check that.
> 
> Having said that, we should probably revisit that. I mitigate the headache around these concerns a bit, because i only use custom plugins with `@react-page/create-plugin-materialui` and non of the built-in plugins. Would be cool to migrate the built-in plugins to use that (or a more general) factory-function.
> 
> EDIT: actually, it already works now, the `<Control />` component of the background plugin did not pass anything special to the Renderer unlike the image plugin.
> 
> I now will check the `Could not find editable: undefined` error

Yeah I know, it's a little bit messy right now. I am actually responsible for some of that code. Needles to say, it was sewed with a hot needle :) Would be great if this all could be unified and improved by using a proper pattern for initial configuration. Most of our plugins are custom as well but we still use some from the lib.



> @PeterKottas i could not reproduce `Could not find editable: undefined` but i think i know where it could come from... and i think its already fixed.
> 
> can you tell me which version did you use?
> 
> 0.16.0-beta.7 is been baked now and should fix all issues hopefully

Great, I'll check it out right away! 

One extra thing that came up while upgrading. Would you pls be able to add a theme prop to the BottomToolbar? Then on 
https://github.com/react-page/react-page/blob/master/packages/ui/src/BottomToolbar/index.tsx#L52
we could do something like:
```
theme={this.props.theme ? this.props.theme : dark ? darkTheme : null}
```
The reason for it is right now, it's not possible to override the theme for this new toolbar. I think this whole thing with themes and how they are handled might need some work in the future as well.

**Comment by PeterKottas** [comment]
Awesome, the new version fixes most of the things. One thing I had to do on my end is add 
```
{...props}
```
To <BottomToolbar> when rendering. It appears that's caused by the editable being a required prop because it is used in the duplicate and draft button. Weird thing though it's not picked up by typescript typings. It doesn't give me the warning despite the prop clearly being required.

One last thing is I now constantly get a warning:
```
Material-UI: you cannot use a custom insertionPoint and <StylesContext injectFirst> at the same time.
```
Not sure if this is on my end or if you see it too.

**Comment by PeterKottas** [comment]
And one small style thing:
This https://github.com/react-page/react-page/blob/0.16.0-beta.7/packages/ui/src/DraftSwitch/index.tsx#L24 and 2 lines bellow, the margin should be removed and marginBottom: 0 should be added here : https://github.com/react-page/react-page/blob/0.16.0-beta.7/packages/ui/src/DraftSwitch/index.tsx#L11 This will center things quite nicely. Right now it's a bit out of balance.

**Comment by macrozone** [comment]
@PeterKottas ok, i will see what i can do about the styling.

I already struggled a bit with the toolbar and its styling. I think you can Wrap it in a ThemeProvider i guess? I defaulted it to the light theme, because most ui elements worked a bit better in bright than in dark theme. Only slate defaults to dark. 

> `{...props}`
> To when rendering. It appears that's caused by the editable being a required prop because it is used in the duplicate and draft button. Weird thing though it's not picked up by typescript typings. It doesn't give me the warning despite the prop clearly being required.

yes, exactly! the Bottom Toolbar now needs to know the id of the current Cell and the editable name (which is only used when having multiple editables in one editor). This is not very clean anymore, so i would suggest to:

- Rename the BottomToolbar to something like NodeActionToolbar, because it already comes with some basic actions (delete, draft, copy)
- Only pass the id (and editable) to that and use connect to get the cell properties you need. (E.g. in de DraftButton i already do this)

Speaking about.. i would love to migrate to redux hooks (or even something like the lovely https://github.com/ctrlplusb/easy-peasy redux  library) and/or use `React.createContext` more often (which is so nice with `useContext`-hook. Then we could e.g. provide the current cell information through the context.

But for the moment, i tried to keep the current style to not make it too messy... but i think we should start migrate some stuff in the near future, because i think it would make it much more readable and understandable and help evolve it further

**Comment by PeterKottas** [comment]
> I already struggled a bit with the toolbar and its styling. I think you can Wrap it in a ThemeProvider i guess? I defaulted it to the light theme, because most ui elements worked a bit better in bright than in dark theme. Only slate defaults to dark.

This unfortunately doesn't work because the bottomToolbar renders it's own theme provider here: https://github.com/react-page/react-page/blob/master/packages/ui/src/BottomToolbar/index.tsx#L52 making it impossible to override the "NodeActionToolbar" if you will. You can only render it inside and override for your own components. Not good. IMHO, it shouldn't render the theme provider at all! Tbh, I think the whole theme handling should be removed from react-page. The fact that we use Material-UI is kind of a choice that was made, not a necessity for the lib itself. If anybody commits to use the controls components provided by the lib, they should be aware of the fact the Material-UI lib is used and the theme should be provided from outside. Right now, it's a little bit messy that we provide some defaults and then give a way of overriding that actually doesn't always work. I would love to hear what you think about it as this would obviously be a breaking change. But maybe one that should be considered.

Yeah, new name would for sure explain better what this does. On the other hand, people who use it will probably complain about the breaking change, and people who don't use it, they wouldn't even notice. Maybe this one is not worth the fuss. As long as typescript works properly with expected props that it requires.

> Speaking about.. i would love to migrate to redux hooks (or even something like the lovely https://github.com/ctrlplusb/easy-peasy redux library) and/or use React.createContext more often (which is so nice with useContext-hook. Then we could e.g. provide the current cell information through the context.

I think something should definitely be done about this at some point. I wasn't sure that redux in general was the best choice for a long time. Obviously it made sense at the time, but I do agree that there are now better options.

**Comment by PeterKottas** [comment]
Haven't seen https://github.com/ctrlplusb/easy-peasy before. Lovely little lib! I would probably need a closer look just to make sure it works with middleware, thinks and so on. But the low configuration complexity as opposed to what redux has is tempting.

**Comment by macrozone** [comment]
> bh, I think the whole theme handling should be removed from react-page. The fact that we use Material-UI is kind of a choice that was made, not a necessity for the lib itself. If anybody commits to use the controls components provided by the lib, they should be aware of the fact the Material-UI lib is used and the theme should be provided from outside.

I disagree on that. Abstracting a UI away is very hard. Using material-ui is of course opinionated, but its neutral and customizable and therefore a good solution. 

Since we lazy load it, its bloat is no longer an argument. I don't see much business-value of themeing or even replace the UI. E.g. take wordpress, it has a default ui on the backend and does not enforce theming it. I even think that having a neutral ui makes it easier to separate it from the _content_, which is of course on-brand. 

But the main reason for me is that if we need to support every possible customization, we hinder innovation and development. Take my additions above with the delete, copy and draft-button. If I had to also support changing the appearance and icons of these buttons, i would never have finished that feature. As initially mentioned, that is kind of a hard problem. Of course, limiting it to themeing in material-ui mitigates that problem, but its still a debt that you have to fullfill.

So bottom line: We should make the default UI as good as possible to fullfill the tasks that editors want to do. that should be the top priority. Devs are invited to raise issues and provide PRs to make it even better, and add customization options to improve it for some use cases. E.g. themeing could make sense when dealing with special content that look odd with the default theme or for special screen sizes.

**Comment by macrozone** [comment]
> And one small style thing:
> This https://github.com/react-page/react-page/blob/0.16.0-beta.7/packages/ui/src/DraftSwitch/index.tsx#L24 and 2 lines bellow, the margin should be removed and marginBottom: 0 should be added here : https://github.com/react-page/react-page/blob/0.16.0-beta.7/packages/ui/src/DraftSwitch/index.tsx#L11 This will center things quite nicely. Right now it's a bit out of balance.
> This in addition to making the theme injectable into the bottom toolbar will make things absolutely beautiful! :)

I look into that, but:

- FormControlLabel has no margin bottom by default in the default theme, so adding marginBottom: 0 there makes no sense to me
- removing `{ marginTop: 5 }` will make the icons non-centered in the default theme, but i have to admit that the adding a top margin is not the best solution there.

But as you can see, that supports my concern about theming. 

I will add a theme property to the BottomToolbar so you can check if this fixes your issue with the non-centered icons. If not, can you open an issue with a screenshot  and the involved browser? maybe its browser specific. and check if you can reproduce it with the example

**Comment by PeterKottas** [comment]
> > bh, I think the whole theme handling should be removed from react-page. The fact that we use Material-UI is kind of a choice that was made, not a necessity for the lib itself. If anybody commits to use the controls components provided by the lib, they should be aware of the fact the Material-UI lib is used and the theme should be provided from outside.
> 
> I disagree on that. Abstracting a UI away is very hard. Using material-ui is of course opinionated, but its neutral and customizable and therefore a good solution.
> 
> Since we lazy load it, its bloat is no longer an argument. I don't see much business-value of themeing or even replace the UI. E.g. take wordpress, it has a default ui on the backend and does not enforce theming it. I even think that having a neutral ui makes it easier to separate it from the _content_, which is of course on-brand.
> 
> But the main reason for me is that if we need to support every possible customization, we hinder innovation and development. Take my additions above with the delete, copy and draft-button. If I had to also support changing the appearance and icons of these buttons, i would never have finished that feature. As initially mentioned, that is kind of a hard problem. Of course, limiting it to themeing in material-ui mitigates that problem, but its still a debt that you have to fullfill.
> 
> So bottom line: We should make the default UI as good as possible to fullfill the tasks that editors want to do. that should be the top priority. Devs are invited to raise issues and provide PRs to make it even better, and add customization options to improve it for some use cases. E.g. themeing could make sense when dealing with special content that look odd with the default theme or for special screen sizes.
> 
> Edit: @PeterKottas what is the main reason why you want to theme the editor?

I think you misunderstood what I meant. What I am saying is we might consider removing just the ThemeProvider. As people use their own theme provider in their app and it's fair to assume they will want to continue using that for all parts of their app. 

I definitely didn't mean abstracting away the whole UI or adding the ability to replace icons or anything like that.

TBH, I'm not too pushy on this, it was just a though. What I insist on though is we should have a way to override all themed material-ui components from outside of react-page. This is not the case now for the new toolbar. It works everywhere else though.

To answer your last question, being able to override the theme is (in general) extremely important. Take the fuss about the dark/light team that the whole web community is talking about now. Also right now, we have a fairly concise, well-designed UI with our company palette, but when I go to the content editor, there is this bottom toolbar that has one pink delete button, purple switch and another purple button. Colors that are not present anywhere else on our dashboard. This problem can be fixed very easily, all it needs is adding a theme prop on the BottomToolbar and then here https://github.com/react-page/react-page/blob/master/packages/ui/src/BottomToolbar/index.tsx#L52, instead of null passing along that new theme props thus making in overridable.

Generally speaking, no matter how "well" you design the UI, there will always be people who want to change a thing or 2 because it's all subjective. That's why themes exist. It would be crazy to allow for overriding of all components, but theme should be the bare minimum. Especially when material-ui supports it quite nicely :)

**Comment by PeterKottas** [comment]
> 
> I look into that, but:
> 
> * FormControlLabel has no margin bottom by default in the default theme, so adding marginBottom: 0 there makes no sense to me
> * removing `{ marginTop: 5 }` will make the icons non-centered in the default theme, but i have to admit that the adding a top margin is not the best solution there.
> 
> But as you can see, that supports my concern about theming.
> 
> I will add a theme property to the BottomToolbar so you can check if this fixes your issue with the non-centered icons. If not, can you open an issue with a screenshot and the involved browser? maybe its browser specific. and check if you can reproduce it with the example

Cheers, I'll have to pull down the code and look for myself in the lib. We should definitely be able to center the icons without the use of any magic margins.

**Comment by macrozone** [comment]
@PeterKottas unfortuatly, the deploy fails constantly on travis, i have no clue why. All tests are passing, so  i could not release it yet. 

i see the point with the themeprovider and i try to leaverage it as much as possible, but i have too little experience and i don't think its _that_ important, that's why i would  love to see others help on that topic!

**Comment by macrozone** [comment]
> > I look into that, but:
> > 
> > * FormControlLabel has no margin bottom by default in the default theme, so adding marginBottom: 0 there makes no sense to me
> > * removing `{ marginTop: 5 }` will make the icons non-centered in the default theme, but i have to admit that the adding a top margin is not the best solution there.
> > 
> > But as you can see, that supports my concern about theming.
> > I will add a theme property to the BottomToolbar so you can check if this fixes your issue with the non-centered icons. If not, can you open an issue with a screenshot and the involved browser? maybe its browser specific. and check if you can reproduce it with the example
> 
> Cheers, I'll have to pull down the code and look for myself in the lib. We should definitely be able to center the icons without the use of any magic margins.

👍

**Comment by PeterKottas** [comment]
My bad, the 5 bottom margin was in my reboot for some reason! All good now.

**Comment by PeterKottas** [comment]
> @PeterKottas unfortuatly, the deploy fails constantly on travis, i have no clue why. All tests are passing, so i could not release it yet.
> 
> i see the point with the themeprovider and i try to leaverage it as much as possible, but i have too little experience and i don't think its _that_ important, that's why i would love to see others help on that topic!

No worries, I have some more experience here so I can help when necessary. Have to set up the code as well. I tried the other day but I think there are some problems with it now on Windows. I think people committed some changes specific for Mac in build scripts. I'll have to look at that.

**Comment by macrozone** [comment]
@PeterKottas version 0.16.0-beta.9 should contain the changes above, but the meta package has not been published, but all others should work

**Comment by PeterKottas** [comment]
Already pulled down and works like a charm! Cheers
